bobrun/wrap.go

```diff
@@ -0,0 +1,236 @@
+	defer errz.Recover(&err)
+
+	rw := &RunWrapper{
+		inner: rc,
+		run:   r,
+		ctx:   ctx,
+
+		done: make(chan struct{}),
+	}
+
+	// create pipes for stdout, stderr and stdin
+	rw.stdout.r, rw.stdout.w, err = os.Pipe()
+	if err != nil {
+		return nil, err
+	}
+
+	rw.stderr.r, rw.stderr.w, err = os.Pipe()
+	if err != nil {
+		return nil, err
+	}
+
+	// react to done from inner control
+	go func() {
+		<-rw.inner.Done()
+		<-rw.initDone
+		close(rw.done)
+	}()
+
+	return rw, nil
+}
+
+func (rw *RunWrapper) Name() string {
+	return rw.inner.Name()
+}
+
+func (rw *RunWrapper) Restart() (err error) {
+	return rw.inner.Restart()
+}
+
+func (rw *RunWrapper) Start() (err error) {
+	defer errz.Recover(&err)
+
+	err = rw.inner.Start()
+	errz.Fatal(err)
+
+	// Wait for initial command to have started
+	for !rw.inner.Running() {
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	boblog.Log.Info(fmt.Sprintf("Init [%s] ", rw.inner.Name()))
+
+	return rw.init()
+}
+
+func (rw *RunWrapper) Stop() error {
+	rw.mux.Lock()
+	if rw.initRunning {
+		rw.initCtxCancel()
+	}
+	rw.mux.Unlock()
+
+	// wait for init to shutdown or deadline is reached.
+	select {
+	case <-rw.initDone:
+	case <-time.After(5 * time.Second):
+	}
+
+	return rw.inner.Stop()
+}
+
+func (rw *RunWrapper) Running() bool {
+	return rw.inner.Running()
+}
+
+func (rw *RunWrapper) Shutdown() error {
+	rw.mux.Lock()
+	if rw.initRunning {
+		rw.initCtxCancel()
+	}
+	rw.mux.Unlock()
+
+	return rw.inner.Shutdown()
+}
+
+func (rw *RunWrapper) Done() <-chan struct{} {
+	return rw.done
+}
+
+func (rw *RunWrapper) Stdout() io.Reader {
+	return io.MultiReader(rw.inner.Stdout(), rw.stdout.r)
+}
+func (rw *RunWrapper) Stderr() io.Reader {
+	return io.MultiReader(rw.inner.Stderr(), rw.stderr.r)
+}
+func (rw *RunWrapper) Stdin() io.Writer {
+	return rw.inner.Stdin()
+}
+
+func (rw *RunWrapper) init() (err error) {
+	rw.mux.Lock()
+	defer rw.mux.Unlock()
+
+	if rw.initRunning {
+		return nil
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	rw.initCtxCancel = cancel
+	rw.initRunning = true
+	rw.initDone = make(chan struct{})
+
+	// React to cancel from external context
+	go func() {
+		select {
+		case <-rw.ctx.Done():
+		case <-rw.initDone: // assures this goroutine exits correctly
+		}
+		rw.initCtxCancel()
+	}()
+
+	go func() {
+		for _, run := range rw.run.init {
+			p, err := syntax.NewParser().Parse(strings.NewReader(run), "")
+			if err != nil {
+				boblog.Log.V(1).Error(err, "shell command parse error")
+				break
+			}
+
+			// FIXME: make run cmds ready for nix integration.
+			env := os.Environ()
+
+			pr, pw, err := os.Pipe()
+			if err != nil {
+				boblog.Log.V(1).Error(err, "creating a pipe")
+				break
+			}
+
+			s := bufio.NewScanner(pr)
+			s.Split(bufio.ScanLines)
+
+			go func() {
+				for s.Scan() {
+					err := s.Err()
+					if err != nil {
+						return
+					}
+
+					// FIXME: why does printing to Commands stdout not work?
```

Comment:
@rdnt i have several FIXME's related to piping output to the status tab.. Does it make sense to look over them in this PR, or shall we create a entire Issue to look into the topic (i know you already did a lot to make it work).